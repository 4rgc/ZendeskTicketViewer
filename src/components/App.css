.App {
	text-align: center;
	display: flex;
	flex-direction: column;
	height: 100%;
	background-color: #001514;
}

.App-link {
	color: #61dafb;
}

.table-container {
	flex: 1;
	overflow-y: scroll;
	padding: 20px;
	align-items: center;
}
