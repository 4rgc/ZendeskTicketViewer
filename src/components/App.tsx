import { MouseEvent, useEffect, useState } from 'react';
import { Column } from 'react-table';
import useSWR from 'swr';
import { ZendeskTicket } from '../types/ZendeskTicket';
import './App.css';
import Header from './Header/Header';
import Table from './Table/Table';
import { TicketDetails } from './TicketDetails/TicketDetails';
import _ from 'lodash';
import moment from 'moment';
import { ErrorView } from './ErrorView/ErrorView';
import { Paging } from './Paging/Paging';
import { fetcher, FetcherError } from '../util/fetcher';

type APIResponse = {
	tickets?: ZendeskTicket[];
	totalCount?: number;
	error?: string;
};

function App() {
	const apiUrl =
		'http://ec2-35-183-81-115.ca-central-1.compute.amazonaws.com:8080/tickets?limit=25&page=';

	const [displayData, setDisplayData] = useState<APIResponse['tickets']>();
	const [apiError, setApiError] = useState<string>('');
	const [page, setPage] = useState<number>(0);
	const [pageCount, setPageCount] = useState<number>(0);
	const [selectedTicket, setSelectedTicket] = useState<ZendeskTicket | null>(
		null
	);
	const { data, error } = useSWR<APIResponse, FetcherError>(
		`${apiUrl}${page}`,
		fetcher
	);
	const perPage = 25;

	useEffect(() => {
		if (data) {
			const tickets = data.tickets;
			if (tickets) {
				const clonedTickets = _.cloneDeep(tickets);
				clonedTickets.forEach((ticket) => {
					ticket.created_at = moment(ticket.created_at).calendar();
					ticket.updated_at = moment(ticket.updated_at).calendar();
				});
				setDisplayData(clonedTickets);
				if (data.totalCount)
					setPageCount(Math.ceil(data.totalCount / perPage));
			}
			if (data.error) {
				setApiError(data.error);
			} else {
				setApiError('');
			}
		}
	}, [data]);

	useEffect(() => {
		if (error) {
			setApiError(
				`${error.status}: ${error.message}. ${error.info?.error}`
			);
		} else {
			setApiError('');
		}
	}, [error]);

	const tableHeaders: Column<Partial<ZendeskTicket>>[] = [
		{
			Header: 'Subject',
			accessor: 'subject',
		},
		{
			Header: 'Status',
			accessor: 'status',
		},
		{
			Header: 'Created At',
			accessor: 'created_at',
		},
		{
			Header: 'Updated At',
			accessor: 'updated_at',
		},
	];

	const onRowClick = (
		event: MouseEvent<HTMLTableRowElement>,
		ticket: ZendeskTicket
	) => {
		setSelectedTicket(ticket);
	};

	const onTicketDetailsClose = () => {
		setSelectedTicket(null);
	};

	const onPageSelected = (page: number) => setPage(page);
	const onPrevPageClicked = () => setPage(page - 1);
	const onNextPageClicked = () => setPage(page + 1);

	return (
		<div className="App">
			<Header />
			{apiError !== '' ? (
				<ErrorView type="error" message={apiError} />
			) : null}
			<div className="table-container">
				<Table
					data={displayData ? displayData : []}
					columns={tableHeaders}
					onRowClick={onRowClick}
				/>
			</div>
			{!apiError ? (
				<Paging
					size="small"
					page={page}
					pageCount={pageCount}
					showPrevNext
					onPageSelected={onPageSelected}
					onPrevSelected={onPrevPageClicked}
					onNextSelected={onNextPageClicked}
				/>
			) : null}
			{selectedTicket && (
				<TicketDetails
					ticket={selectedTicket}
					onClose={onTicketDetailsClose}
				/>
			)}
		</div>
	);
}

export default App;
