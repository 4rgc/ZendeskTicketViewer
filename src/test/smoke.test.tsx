import App from '../components/App';
import renderer from 'react-test-renderer';

describe('App smoke test', () => {
	it('opens up', () => {
		const component = renderer.create(<App />);

		let tree = component.toJSON();

		expect(tree).toMatchSnapshot();
	});
});
