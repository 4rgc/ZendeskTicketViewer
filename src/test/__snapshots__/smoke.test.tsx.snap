// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App smoke test opens up 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <div
      className="header-text"
    >
      Great ticket viewer
    </div>
  </header>
  <div
    className="table-container"
  >
    <table
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colSpan={1}
            role="columnheader"
          >
            Subject
          </th>
          <th
            colSpan={1}
            role="columnheader"
          >
            Status
          </th>
          <th
            colSpan={1}
            role="columnheader"
          >
            Created At
          </th>
          <th
            colSpan={1}
            role="columnheader"
          >
            Updated At
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      />
    </table>
  </div>
  <div
    className="paging-cont"
  >
    <button
      className="btn-base btn-prm btn-sq paging-btn"
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      className="btn-base btn-prm btn-sq paging-btn"
      onClick={[Function]}
    >
      &gt;
    </button>
  </div>
</div>
`;
